### PR TITLE
[AIRFLOW-5477] Rewrite Google PubSub Hook to Google Cloud Python

### PR DESCRIPTION
**Pull request by TobKed**
Make sure you have checked _all_ steps below.

### Jira

- [ ] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-5477
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.
  - In case you are proposing a fundamental code change, you need to create an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)).
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Description

- [ ] Here are some details about my PR, including screenshots of any UI changes:

### Tests

- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [ ] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### CODE REVIEW

File: airflow/gcp/operators/pubsub.py [hunk 1]
```diff
@@ -186,6 +264,13 @@ def __init__(
             fail_if_exists: bool = False,
             gcp_conn_id: str = 'google_cloud_default',
             delegate_to: Optional[str] = None,
+            push_config: Optional[Union[Dict, types.PushConfig]] = None,  # pylint: disable=no-member
+            retain_acked_messages: Optional[bool] = None,
+            message_retention_duration: Optional[Union[Dict, types.Duration]] = None,  # noqa E501 # pylint: disable=no-member,line-too-long
```
Comment (nuclearpinguin) [inline airflow/gcp/operators/pubsub.py]: I think that it would look better to import `PushConfig`, `Duration` from `types` and put pylint disable once.

**Comment by TobKed** [comment]
cc @potiuk @mik-laj

File: airflow/gcp/hooks/pubsub.py [hunk 2]
```diff
@@ -77,47 +103,106 @@ def publish(self, project: str, topic: str, messages: List[Dict]) -> None:
...
+
+        self.log.info("Published %d messages to topic (path) %s", len(messages), topic_path)
 
-    def create_topic(self, project: str, topic: str, fail_if_exists: bool = False) -> None:
+    def create_topic(
+        self,
+        project: str,
+        topic: str,
+        fail_if_exists: bool = False,
+        labels: Optional[Dict[str, str]] = None,
```
Comment (nuclearpinguin) [inline airflow/gcp/hooks/pubsub.py]: According to integration guide we should add label with Airflow version to indicate that a resource was created with the use of Airflow. For example:
https://github.com/PolideaInternal/airflow/blob/6a66ece8cae8c8d209efefcde2639dca8919e975/airflow/gcp/hooks/gcs.py#L449-L451

To obtain version use `from airflow.version import version` ;)

File: airflow/gcp/hooks/pubsub.py [hunk 3]
```diff
@@ -52,18 +57,39 @@ class PubSubHook(GoogleCloudBaseHook):
...
-        :rtype: googleapiclient.discovery.Resource
+        :return: Google Cloud Pub/Sub client object.
+        :rtype: google.cloud.pubsub_v1.PublisherClient
         """
-        http_authorized = self._authorize()
-        return build(
-            'pubsub', 'v1', http=http_authorized, cache_discovery=False)
+        if not self._client:
+            self._client = PublisherClient(
+                credentials=self._get_credentials(),
```
Comment (mik-laj) [inline airflow/gcp/hooks/pubsub.py]: Can you add client info here?

File: airflow/gcp/hooks/pubsub.py [hunk 4]
```diff
@@ -52,18 +57,39 @@ class PubSubHook(GoogleCloudBaseHook):
...
+    @cached_property
+    def subscriber_client(self) -> SubscriberClient:
+        """
+        Creates SubscriberClient.
+
+        :return: Google Cloud Pub/Sub client object.
+        :rtype: google.cloud.pubsub_v1.SubscriberClient
+        """
+        return SubscriberClient(
+            credentials=self._get_credentials(),
```
Comment (mik-laj) [inline airflow/gcp/hooks/pubsub.py]: Can you add client_info here?

Comment (TobKed) [reply airflow/gcp/hooks/pubsub.py]: `SubscriberClient` doesn't get `client_info` argument.

File: airflow/gcp/hooks/pubsub.py (hunk 3 above)
Comment (TobKed) [reply airflow/gcp/hooks/pubsub.py]: `PublisherClient ` doesn't get `client_info` argument.

File: airflow/gcp/hooks/pubsub.py (hunk 2 above)
Comment (TobKed) [reply airflow/gcp/hooks/pubsub.py]: Good idea. I will add it.

File: airflow/gcp/hooks/pubsub.py (hunk 4 above)
Comment (TobKed) [reply airflow/gcp/hooks/pubsub.py]: I was wrong. As you showed me, it takes later from kwargs.

File: airflow/gcp/hooks/pubsub.py (hunk 3 above)
Comment (TobKed) [reply airflow/gcp/hooks/pubsub.py]: I was wrong. As you showed me, it takes later from kwargs.

**Comment by nuclearpinguin** [comment]
Closed via https://github.com/apache/airflow/pull/6096

**Pull request closed without merging**